client/extensions/woocommerce/products/product-variation-types-form/index.js

```diff
@@ -0,0 +1,120 @@
+/**
+ * External dependencies
+ */
+import React, { Component, PropTypes } from 'react';
+import i18n from 'i18n-calypso';
+import clone from 'lodash/clone';
+
+/**
+ * Internal dependencies
+ */
+import FormLabel from 'components/forms/form-label';
+import FormTextInput from 'components/forms/form-text-input';
+import Button from 'components/button';
+import TokenField from 'components/token-field';
+
+export default class ProductVariationTypesForm extends Component {
+
+	static propTypes = {
+		product: PropTypes.shape( {
+			id: PropTypes.number.isRequired,
+			name: PropTypes.string.isRequired,
+			type: PropTypes.string.isRequired,
+		} ),
+		variations: PropTypes.arrayOf( PropTypes.shape( {
+			type: PropTypes.string.isRequired,
+			values: PropTypes.arrayOf( PropTypes.string )
+		} ) ),
+	};
+
+	constructor( props ) {
+		super( props );
+
+		this.state = {
+			variations: this.props.variations ? this.props.variations : this.getInitialFields(),
+		};
+
+		this.addVariation = this.addVariation.bind( this );
+		this.renderInputs = this.renderInputs.bind( this );
+	}
+
+	getInitialFields() {
+		return [ this.getNewFields() ];
+	}
+
+	getNewFields() {
+		return {
+			type: '',
+			values: [],
+		};
+	}
+
+	updateType( index, event ) {
+		event.preventDefault();
+		const newValue = event.target.value,
+			updatedVariations = clone( this.state.variations );
+		updatedVariations[ index ] = clone( updatedVariations[ index ] );
+		updatedVariations[ index ].type = newValue;
+		this.setState( { variations: updatedVariations } );
+	}
+
+	updateValues( index, value ) {
+		const updatedVariations = clone( this.state.variations );
+		updatedVariations[ index ] = clone( updatedVariations[ index ] );
+		updatedVariations[ index ].values = value;
+		this.setState( { variations: updatedVariations } );
+	}
+
+	renderInputs( variation, index ) {
+		return (
+			<div key={index} className="product-variation-types-form__fieldset">
+				<FormTextInput
+					placeholder={ i18n.translate( 'Color' ) }
+					value={ variation.type }
+					name="type"
+					onChange={ this.updateType.bind( this, index ) }
+					className="product-variation-types-form__field"
+				/>
+				<TokenField
+					placeholder={ i18n.translate( 'Comma separate these' ) }
+					value={ variation.values }
+					name="values"
+					onChange={ this.updateValues.bind( this, index ) }
+				/>
+			</div>
+		);
+	}
+
+	addVariation( event ) {
+		event.preventDefault();
+		const updatedVariations = this.state.variations.concat( [ this.getNewFields() ] );
```

Comment:
Updated syntax: `const updatedVariations = [ ...this.state.variations, this.getNewFields() ];`